ocean_lib/models/test/test_datatoken_order_both_templates.py

```diff
@@ -6,7 +6,7 @@
 
 import pytest
```

Comment:
### test_dispense_and_order_with_defaults()

We can optimize the defaults for this even better.

Status quo:

```python
    tx = DT.dispense_and_order(
        consumer=consumer_wallet.address,
        service_index=1,
        provider_fees=provider_fees,
        transaction_parameters={"from": publisher_wallet},
    )
```

Optimize into:
```python
provider_fees  = get_mock_provider_fees(..)
order_tx_id = DT.dispense_and_order(provider_fees, {"from": publisher_wallet})
```

Where:
- it returns `order_tx_id`, not `tx`. Like calls to `pay_for_access_service()`.

With defaults set as:
```python
consumer = transaction_parameters["from"].address
service_index = 1
```

### test_buy_DT_and_order()

We should split test function this into two test functions: one with good defaults, and one with nondefaults. 

### [new] test_buy_DT_and_order_with_defaults()

Status quo: 
```python
    args = {
        "consumer": another_consumer_wallet.address,
        "service_index": 1,
        "provider_fees": provider_fees,
        "consume_market_fees": FeeTokenInfo(
            address=consume_fee_address,
            token=DAI.address,
        ),
        "exchange": exchange,
        "transaction_parameters": {"from": publisher_wallet},
    }

    if template_index == 2:
        args["max_base_token_amount"] = to_wei(2.5)
        args["consume_market_swap_fee_amount"] = to_wei(0.001)  # 1e15 => 0.1%
        args["consume_market_swap_fee_address"] = another_consumer_wallet.address

    tx = DT.buy_DT_and_order(**args)
```

Optimize into:
```python
provider_fees  = get_mock_provider_fees(..)
order_tx_id = exchange.buy_DT_and_order(provider_fees, {"from": publisher_wallet})
```

Where:
- the call is to 'exchange` (`Exchange` class), not `DT` (`Datatoken` class). Just like `buy_DT()`. And `exchange` knows which datatoken to use, by definition
- it returns `order_tx_id`, not `tx`. Like calls to `pay_for_access_service()`.

With defaults set as:
```python
consumer = transaction_parameters["from"].address
service_index = 1
consume_market_fees = <zero>
```

### [new] test_buy_DT_and_order_with_non_defaults()

Status quo: (like shown above)

The "args" setup makes it harder to read. Let's remove that. And, it should also return `order_tx_id` like the defaults flow. So, it will look something like:

Optimize into:
```python
provider_fees  = get_mock_provider_fees(..)
consume_market_order_fees = FeeTokenInfo(address=consume_fee_address, token=DAI.address)
consume_market_swap_fees = FeeTokenInfo(to_wei(0.001), address=another_consumer_wallet.address, token=..)
max_base_token_amount = to_wei(2.5)
order_tx_id = exchange.buy_DT_and_order(
        consumer = another_consumer_wallet.address,
        service_index = 1,
        max_base_token_amount = max_base_token_amount,
        provider_fees = provider_fees,
        consume_market_order_fees = consume_market_order_fees,
        consume_market_swap_fees = consume_market_swap_fees,
        transaction_parameters = {"from": publisher_wallet},
)
```

Where:
- both template 1 and 2 should be aware of max_base_token_amount and consume_market_swap_fees (not just template 2 like the status quo)

Reply:
> optimising to: DT.dispense_and_order(provider_fees, {"from": publisher_wallet})
this test uses the consumer_wallet for the consumer address, while the sender of the transaction is publisher_wallet. I think we the consumer is NOT by default the sender of the transaction. Especially because we mimic DatatokenEnterprise here.

Same comment for the exchange test. Even more so, since we have another_consumer_wallet as the consumer.

Ref: moving this class to Exchange and figuring out there whether to do a direct call from DatatokenEnterprise template vs. combo with multiple txs from Datatoken: I can do that, but there is a code smell here. If the max_base_token_amount, consume_market_swap_fee_amount and consume_market_swap_fee_address are not being used with a regular Datatoken (only used in Enterprise), then what is their purpose? Have I made mistakes in replicating the functionality of buy DT and order in the regular template? Is there some transaction I'm missing that would involve these parameters?

Also, ref. returning the tx_id instead of the transaction itself: pay_for_access_service is a higher level function used in ocean assets. For models that do contract calls, we either return the receipt or wrapped models where suitable (i.e. when returning a new Datatoken or tuple of DataNFT, Datatoken, Exchange objects). We should not switch to tx_ids for functions that reside inside models, since receipts are far more rich and usable. Or, if we do, we need to level everything and do a uniform refactor for all (which should probably be another issue)

Reply:
I'm not sure what the best answer is on this. To really know, I'd have to play around with the possibilities more myself.

Given that we aren't yet in sync on this, and it would be really useful to "ratchet up" ocean.py with all the great improvements of this PR, I recommend:

- merge the PR as-is
- copy these comments into a github issue for further discussion / playing with possibilities. (And probably lower priority than some other things)